arc/src/main.c

```diff
@@ -388,6 +388,250 @@ static void handle_glcd(struct zjs_ipm_message* msg)
+{
+    int32_t val1, val2;
+    double result = 0;
+
+    switch (val->type) {
+    case SENSOR_VALUE_TYPE_INT:
+        result = val->val1;
+        break;
+    case SENSOR_VALUE_TYPE_INT_PLUS_MICRO:
+        if (val->val2 == 0) {
+            result = (double)val->val1;
+            break;
+        }
+
+        /* normalize value */
+        if (val->val1 < 0 && val->val2 > 0) {
+            val1 = val->val1 + 1;
+            val2 = val->val2 - 1000000;
+        } else {
+            val1 = val->val1;
+            val2 = val->val2;
+        }
+
+        if (val1 > 0 || (val1 == 0 && val2 > 0)) {
+            result = val1 + (double)val2 * 0.000001;
+        } else if (val1 == 0 && val2 < 0) {
+            result = (double)val2 * (-0.000001);
+        } else {
+            result = val1 + (double)val2 * (-0.000001);
+        }
+        break;
+    case SENSOR_VALUE_TYPE_DOUBLE:
+        result = val->dval;
+        break;
+    default:
+        PRINT("convert_sensor_value: invalid type %d\n", val->type);
+        return 0;
+    }
+
+    return result;
+}
+
+static void process_accel_data(struct device *dev)
+{
+    struct sensor_value val[3];
+    double dval[3];
+
+    if (sensor_channel_get(dev, SENSOR_CHAN_ACCEL_ANY, val) < 0) {
+        PRINT("Cannot read accelerometer channels.\n");
+        return;
+    }
+
+    dval[0] = convert_sensor_value(&val[0]);
+    dval[1] = convert_sensor_value(&val[1]);
+    dval[2] = convert_sensor_value(&val[2]);
+
+    if (dval[0] == 0 && dval[1] == 0 && dval[2] == 0) {
+        // FIXME: BUG? why sometimes it reports 0, 0, 0 on all axis
+        return;
+    }
+
+    // set slope threshold to 0.1G (0.1 * 9.80665 = 4.903325 m/s^2)
+    double threshold = 0.980665;
+    if (ABS(dval[0] - accel_last_value[0]) > threshold ||
+        ABS(dval[1] - accel_last_value[1]) > threshold ||
+        ABS(dval[2] - accel_last_value[2]) > threshold) {
+        union sensor_reading reading;
+        reading.x = dval[0];
+        reading.y = dval[1];
+        reading.z = dval[2];
+        send_sensor_data(SENSOR_CHAN_ACCEL_ANY, reading);
+    }
+
+#ifdef DEBUG_BUILD
+    char buf_x[18], buf_y[18], buf_z[18];
+
+    sensor_value_snprintf(buf_x, sizeof(buf_x), &val[0]);
+    sensor_value_snprintf(buf_y, sizeof(buf_y), &val[1]);
+    sensor_value_snprintf(buf_z, sizeof(buf_z), &val[2]);
+    PRINT("sending accel: X=%s, Y=%s, Z=%s\n", buf_x, buf_y, buf_z);
+#endif
+}
+
+static void process_gyro_data(struct device *dev)
+{
+    struct sensor_value val[3];
+    double dval[3];
+
+    if (sensor_channel_get(dev, SENSOR_CHAN_GYRO_ANY, val) < 0) {
+        PRINT("Cannot read gyroscope channels.\n");
+        return;
+    }
+
+    dval[0] = convert_sensor_value(&val[0]);
+    dval[1] = convert_sensor_value(&val[1]);
+    dval[2] = convert_sensor_value(&val[2]);
+
+    if (ABS(dval[0] - gyro_last_value[0]) > 0 ||
+        ABS(dval[1] - gyro_last_value[1]) > 0 ||
+        ABS(dval[2] - gyro_last_value[2]) > 0) {
```

Comment:
yeah, i was going to add a threshold here intead of 0, but I don't really know what value to put in for now, so i left it as 0.  I'll update it.